TERRAFORM.md

```diff
@@ -0,0 +1,96 @@
+# Terraform Best Practices and Guidelines
+
+General terraform guidelines and best practices
+
+## Table of Contents
+
+- [Table of Contents](#table-of-contents)
+- [Introduction](#introduction)
+- [Code Organization](#code-organization)
+  - [Reusable Module Usage](#reusable-module-usage)
+- [State Management](#state-management)
+  - [Remote State](#remote-state)
+    - [State key](#state-key)
+  - [State Locking](#state-locking)
+  - [State migration](#state-migration)
+  - [Module Guidelines](#module-guidelines)
+- [Code formatting](#code-formatting)
+
+## Introduction
+
+Terraform is a powerful tool for provisioning and managing infrastructure as code. To ensure maintainability, scalability, and security, it's important to follow best practices and guidelines when working with Terraform.
+
+## Code Organization
+
+The high level code organization should look like this
+
+```text
+|- modules/                   # terraform reusable modules for the whole repository
+|- ORGANIZATIONS_OR_ACCOUNT   # e.g. AWS Account name
+|           |- modules        # terraform reusable modules for this organization/account
+|           |- shared         # shared resources
+|           |- STAGING        # resource pro staging
+```
+
+Then it should mirror the hierarchical structure of the infrastructure. Note that each terraform module (project) should be kept small.
```

Comment:
I am unclear about what the "hierachical structure of the infrastructure" means here.
Also, in order to enforce the "keep it small" constraint, should we make the CI complain when a module/project exceeds N lines of code?

Reply:
Here I wanted to be quite generic and don't provide any informations on our stack infrastructure because the guideline repository is public. You will find our stack hierarchy in our confluence pat.

About keeping it small you cannot really define a number of code line, you might have some "small" terraform module with a lots of line if you need to have configuration file, or if you require to have complex policies. So enforcing this by a CI would be quite difficult, I think the enforcement should be more in the review process, where the developer and the review should have some common sense to define when to split the code into several module.
For example on our stack we have a module that define the "system" aka Cloudfront with DNS and WAF, and have another terraform module that define the service itself (k8s namespace, bucket for service, etc).

Reply:
Fair enough about enforcement but then I suggest explaining in more details what "keep it small" means, maybe with examples. Like you did in this review comment.

For the structure, what about a link to Confluence? I understand external readers won't be able to click it but the audience for this documentation explaining how our Terraform config works is internal anyway since external people don't have access to that either.